sqlmesh/core/context.py

```diff
@@ -574,15 +575,38 @@ def load(self, update_schemas: bool = True) -> GenericContext[C]:
             self._requirements.update(project.requirements)
             self._excluded_requirements.update(project.excluded_requirements)
 
+        if any(self._projects):
+            prod = self.state_reader.get_environment(c.PROD)
+
+            if prod:
+                for snapshot in self.state_reader.get_snapshots(prod.snapshots).values():
+                    if snapshot.node.project not in self._projects:
+                        store = self._standalone_audits if snapshot.is_audit else self._models
+                        store[snapshot.name] = snapshot.node  # type: ignore
+
         for model in self._models.values():
             self.dag.add(model.fqn, model.depends_on)
 
-        # This topologically sorts the DAG & caches the result in-memory for later;
-        # we do it here to detect any cycles as early as possible and fail if needed
-        self.dag.sorted
-
         if update_schemas:
+            for fqn in self.dag:
+                model = self._models.get(fqn)  # type: ignore
+
+                if not model or not model._data_hash:
+                    continue
+
+                # make a copy of remote models that depend on local models or in the downstream chain
+                # without this, a SELECT * FROM local will not propogate properly because the downstream
+                # model will get mutated (schema changes) but the object is the same as the remote cache
+                if any(
+                    not self._models[dep]._data_hash
```

Comment:
Does missing `_data_hash` signals that the model wasn't loaded from cache or that it's a local model? If it's the latter can we check by project name instead?

Reply:
project is not enough, you need the tree of things that are invalidated